File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void recur(TreeNode* root, int pathSum, int& totalSum) {
        pathSum = pathSum * 10 + root->val;
        if (root->left == nullptr && root->right == nullptr) {
            totalSum += pathSum;
            return;
        }
        if (root->left) {
            recur(root->left, pathSum, totalSum);
        }
        if (root->right) {
            recur(root->right, pathSum, totalSum);
        }
    }
    int sumNumbers(TreeNode* root) {
        int ans = 0;
        recur(root, 0, ans);
        return ans;
    }
};

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int totalSum;
    
    public int sumNumbers(TreeNode root) {
        totalSum = 0;
        recur(root, 0);
        return totalSum;
    }
    
    public void recur(TreeNode root, int pathSum) {
        if (root == null) {
            return;
        }
        pathSum = pathSum * 10 + root.val;
        if (root.left == null && root.right == null) {
            totalSum += pathSum;
            return;
        }
        recur(root.left, pathSum);
        recur(root.right, pathSum);
    }
}